scheduler/rank.go

```diff
@@ -202,13 +203,21 @@ OUTER:
 		}
 
 		// Index the existing network usage
-		netIdx := structs.NewNetworkIndex()
-		netIdx.SetNode(option.Node)
-		netIdx.AddAllocs(proposed)
+		var netIdx *structs.NetworkIndex
+		initNetIdx := func() {
+			netIdx = structs.NewNetworkIndex()
+			netIdx.SetNode(option.Node)
+			netIdx.AddAllocs(proposed)
+		}
+		var initNetIdxOnce sync.Once
```

Comment:
I like think the functionality is right, but let's rework this not require a closure or a sync.Once. Since netIdx is never shared between goroutines a bool or `if netIdx == nil { init() }` check is sufficient.

However, I want us to at least consider refactoring NetworkIndex instead before merging because like many of networking structs it's used for slightly different purposes in slightly different places.

Here NetworkIndex should _lazily_ evaluate SetNode+AddAllocs only when AssignPorts is called.

The collision detection logic that requires immediate (non-lazy) evaluation is only used in the planner.

Perhaps SetNode+AddAllocs could be made lazy (just setters that copy the pointer they're passed; probably inlined by the compiler), and a new `Check() (ok bool)` method could be added?


At any rate, this is great work and the refactorings may not pan out, but I think pushing a bit further on it could make it more readable and perhaps even a teeny tiny bit more efficient? That may not be true since I think in the no-networks path your approach will incur 0 heap allocations.